### PR TITLE
Initialize Driver API before calling it in Cython

### PR DESCRIPTION
**Pull request by jakirkham**
Make sure that we have called `cuInit` to initialize the CUDA Driver API before making any other Driver API calls.

### CODE REVIEW

File: python/cudf/cudf/_cuda/gpu.pyx [hunk 1]
```diff
@@ -23,20 +23,23 @@ from cudf._cuda.gpu cimport underlying_type_attribute as c_attr
...
         return (type(self), (self.status,))
 
 
 class CUDADriverError(RuntimeError):
 
     def __init__(self, CUresult status):
-        self.status = status
+        cdef CUresult init_status = cuInit(0)
+        if init_status != 0:
+            raise CUDADriverError(init_status)
```
Comment (jakirkham) [inline python/cudf/cudf/_cuda/gpu.pyx:42]: So this is probably not a viable way to handle this error as we could wind up in infinite recursion. It is possible `cuInit` would just be kicking up some error caused by an unrelated operation. However more likely if this is failing there is some more fundamental configuration issue. Not sure how we want to handle that. One option would be just to raise a `RuntimeError` with some message and the status. Another might be to check for the few error status here and handle them accordingly. There may be other options.

File: python/cudf/cudf/_cuda/gpu.pyx [hunk 2]
```diff
@@ -367,12 +372,17 @@ def deviceGetName(int device):
     and status code.
     """
 
+    cdef CUresult init_status = cuInit(0)
```
Comment (kkraus14) [inline python/cudf/cudf/_cuda/gpu.pyx:375]: This is setting a bad precedent that we're going to need to add a `cuInit` call in every driver API wrapper we build.

Should we just initialize in an `__init__.py(x)` somewhere?

Comment (jakirkham) [reply python/cudf/cudf/_cuda/gpu.pyx:375]: Apparently `cuInit` only incurs a cost once (based on asking internally).

That said, I don't have strong feelings on how we do this. Though it does appear we do need to do this for the Driver API calls to work correctly.

Comment (kkraus14) [reply python/cudf/cudf/_cuda/gpu.pyx:375]: Couldn't we just call a `cuInit` within the Cython module `__init__.pyx` file to guarantee the driver is initialized before any function can be used?

Comment (jakirkham) [reply python/cudf/cudf/_cuda/gpu.pyx:375]: I'd rather not slow down `import` time further. Outside of that I don't really have any preferences.

Comment (kkraus14) [reply python/cudf/cudf/_cuda/gpu.pyx:375]: > I'd rather not slow down `import` time further. Outside of that I don't really have any preferences.

We're already calling runtime APIs on import that implicitly initialize the driver (without creating a context), so this shouldn't change import time.

**Comment by harrism** [comment]
Closing  as stale.

**Comment by jakirkham** [comment]
Yeah I think if we want to do this we would do it in RMM since that's where this module moved to.

xref: https://github.com/rapidsai/rmm/pull/489
xref: https://github.com/rapidsai/cudf/pull/6012

**Pull request closed without merging**